exploration/0006-number-selection.md

```diff
@@ -0,0 +1,127 @@
+# Selection on Numerical Values
+
+Status: **Proposed**
+
+<details>
+	<summary>Metadata</summary>
+	<dl>
+		<dt>Contributors</dt>
+		<dd>@eemeli</dd>
+		<dt>First proposed</dt>
+		<dd>2023-09-06</dd>
+		<dt>Pull Request</dt>
+		<dd>#000</dd>
+	</dl>
+</details>
+
+## Objective
+
+Define how selection on numbers happens.
+
+## Background
+
+As discussed on numerous past occasions and explicitly identified in
+<a href="https://github.com/unicode-org/message-format-wg/pull/457">#457</a>,
+it would be good for both plural and non-plural match selection to be supported in MF2.
+
+MF1 supported selection on either cardinal plurals or ordinal numbers,
+via the `plural` and `selectordinal` selectors.
+It also customized this selection beyond the capabilities of `com.ibm.icu.text.PluralRules`
+by allowing for an `offset` parameter.
+
+As pointed out by <a href="https://github.com/mihnita">@mihnita</a> in particular,
+category selection is not always appropriate for selection on a number:
+the number may be representing some completely other quantity,
+such as a four-digit year or the integer value of an enumerator.
+
+Furthermore, as pointed out by <a href="https://github.com/ryzokuken">@ryzokuken</a>
+in <a href="https://github.com/unicode-org/message-format-wg/pull/457#discussion_r1307443288">#457 (comment)</a>,
+ordinal values are not truly "plural",
+even if they use categories with similar names and are determined by similar rules.
+
+## Use-Cases
+
+Put together, we have thus identified three different types of selection
+we would like to support on numeric values:
+
+- cardinal plural selection
+- ordinal selection
+- exact match selection
+
+## Requirements
+
+- Enable cardinal plural selection.
+- Enable ordinal number selection.
+- Enable exact match selection.
+- Support relevant formatting options, such as `minimumFractionDigits`.
+- Encourage developers to use the same options for formatting and selection.
+
+## Constraints
+
+ICU MF1 messages using `plural` and `selectordinal` should be representable in MF2.
+
+## Proposed Design
+
+Given that we already have a `:number`,
+it makes sense to add a `<matchSignature>` to it with an option
```

Comment:
Actually... we don't yet have a `:number`. I kind of expected to find it in this doc! A complete definition would help figure out if this is a good design.

My first reaction to your proposal to use `:number` for both was that it was conceptually complicated (functions do multiple _different_ things depending on context), but I can see how it could be potentially both a simplifier and thus more robust for users who don't understand plurals. 

At the same time, I think that some common shorthands might be useful, just as MF1 has `{count,number,integer}`.

I also am concerned about the need to do multiple annotation (this has to do with the mutability question we're working), e.g.:

```
let $count = {$count :number groupingUsed=true 
               minimumFractionDigits=0 minimumDecimalDigits=4 
               maximumFractionDigits=2} <- formatting
match {$count :number type=plural} <- get the right selector
when ...
```

I recognize that my example uses the default selector `plural`, and so wouldn't need `type=`, but how do I get the `exact` selector if I want it without hosing all of that formatting?

Reply:
> Actually... we don't yet have a `:number`. I kind of expected to find it in this doc! A complete definition would help figure out if this is a good design.

We have its definition in [`spec/registry.xml`](https://github.com/unicode-org/message-format-wg/blob/main/spec/registry.xml), which includes a `<formatSignature>` for it. Follow-up work from #457 should probably add a document describing that and `:datetime` to start with.

> At the same time, I think that some common shorthands might be useful, just as MF1 has `{count,number,integer}`.

Possibly yes, but this sounds like a separate follow-on spec PR or design doc. :)

> I also am concerned about the need to do multiple annotation (this has to do with the mutability question we're working), e.g.:
> 
> ```
> let $count = {$count :number groupingUsed=true 
>                minimumFractionDigits=0 minimumDecimalDigits=4 
>                maximumFractionDigits=2} <- formatting
> match {$count :number type=plural} <- get the right selector
> when ...
> ```
> 
> I recognize that my example uses the default selector `plural`, and so wouldn't need `type=`, but how do I get the `exact` selector if I want it without hosing all of that formatting?

First of all, the `:number` in your selector would be getting the resolved rather than formatted value of the first expression, so it could extend rather than clobber the set of options defined in the variable declaration. So just using `type=exact` where you have `type=plural` should do the trick.

Alternatively, you should be able to include it in the first bag of options:
```
let $count = {$count :number type=exact groupingUsed=true 
               minimumFractionDigits=0 minimumDecimalDigits=4 
               maximumFractionDigits=2}
match {$count}
when ...
```

If/as we'll want to keep the formatting and selection options compatible, you should be able to assign a bag of options covering both to a variable.